### ISSUE TITLE
Docker Stack Logs

### ISSUE THREAD

**Issue by mark-church** [docker/docker#31458]
**Description**

This is a request for a `docker stack logs` which can show the logs for a `docker stack` much like `docker service logs` work in 1.13. docker-compose works similarly today, showing the interleaved logs for all containers deployed from a compose file. This will be useful for troubleshooting any kind of errors that span across heterogeneous services.

cc @dperny

**Comment by chawlanikhil24** [comment]
Can  I also participate in contribution to this feature ?

**Comment by thaJeztah** [comment]
@chawlanikhil24 thanks, I'll let @dperny answer that, because I know he's working on changes in this area.

**Comment by dperny** [comment]
@chawlanikhil24 there are some changes that have to be made to the API before we can pursue this, because right now we can only get the logs for 1 service at a time unless you make multiple calls (which is silly, because we can get the logs for multiple services in the same stream on swarmkit's side).

After I finish those API changes, this can be done entirely on the client side, and should be really straightforward. I don't know when the API changes will be in because I have started yet, but I can let you know as soon as I have them!

**Comment by chawlanikhil24** [comment]
@dperny , Thank you much for the information and letting me to pursue this issue. I'll be waiting for your response. Please let me know whenever you are done. @thaJeztah , Thank you too, for the help. :)

**Comment by chawlanikhil24** [comment]
@dperny , Is there any progress has been made on this issue? I am waiting for my part to get initiated.

**Comment by dperny** [comment]
Hey, @chawlanikhil24 

Thanks for reminding me! We didn't get the API endpoint I wanted in, so it's not possible to stream multiple services at the same time. The only option, then, is to do multiple logs calls and combine the results on the client side, which is kind of a mess and a lot more difficult. You're welcome to take a stab at it if you want but I can't make any promises about whether people will think that's right or not. 

IIRC you can identify services belonging to a stack because they're labeled, so you can do a service list call with a label filter of the stack name or ID. Then, you would do a logs call for all of the services. Finally, you would combine all of the responses into one stream, line-by-line. I feel like the last bit might be trickier than it seems.

Thanks for taking a look at this! Sorry it's panning out to be not as simple as I'd hoped.

**Comment by chawlanikhil24** [comment]
Hi @dperny,

Thank you very much for the instructions. I'll start doing my part ASAP. And can I contact you on IRC or any other channel for directions, when needed the most.

**Comment by dperny** [comment]
Hey, @chawlanikhil24,

Sorry about the late reply! Dockercon and all that!

I believe my email address is on my github profile. I'm also on the Docker Community slack. I don't check it regularly but if you ping me directly, I will get the notification. I cannot be reached by IRC. I'm happy to help in whatever way you need it! If you weren't doing this, I'd be doing it, so it's not like you can take up too much of my time on it.

**Comment by knutole** [comment]
Is there another way of receiving logs from `docker stack deploy` in the meantime? 

I'm getting `only supported with experimental daemon` when doing `docker services logs -f mapic`.


Docker version:
```
Client:
 Version:      17.03.1-ce
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Mon Mar 27 17:10:36 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.1-ce
 API version:  1.27 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   c6d412e
 Built:        Mon Mar 27 17:10:36 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Comment by dperny** [comment]
@knutole service logs came out of experimental in 17.05. You can enable experimental features by passing the `--experimental` flag when you start the docker daemon. 

If you want to get stack logs in the mean time, my best offer is a shell one-liner that does `docker stack services -q STACKNAME` and then launches and detaches (with `&`) service logs on each service.

**Comment by Wolfium** [comment]
As a possible workaround in the meantime...

I use very often something that I discovered by mistake.

you can use substrings names from docker service ps, mainly if you pass only stack name, it will show all services within the stack and you pass a stack name plus underscore (default contact for stacks) plus service name within stack you get only those services instances details.
More over it works with auto completation indeed.

Example stack name show is "pe"
```
# docker service ps pe
ID                  NAME                IMAGE                                                           NODE                       DESIRED STATE       CURRENT STATE               ERROR                              PORTS
reonqbue12wk        pe_backend.1        registry:develop        gtiDesaUbuDockerWorker01   Running             Running 6 days ago
u3tdzcxvh27z        pe_web.1            registry:develop-app    gtiDesaUbuDockerWorker01   Running             Running 6 days ago
p0ig9aoiljju         \_ pe_web.1        registry:develop-app    gtiDesaUbuDockerWorker01   Shutdown            Shutdown 6 days ago
k7f6bj7iou7f        pe_backend.1        registry:develop                                   Shutdown            Pending 5 hours ago
dw5wwqyy02zx         \_ pe_backend.1    registry:develop                                   Shutdown            Pending 5 hours ago
uctb1gcsrkdl        pe_backend.2        registry:develop        gtiDesaUbuDockerWorker03   Running             Running 6 days ago
t878f2d9deny         \_ pe_backend.2    registry:develop        gtiDesaUbuDockerWorker02   Shutdown            Rejected 6 days ago         "No such image:"
tvlc28rq45ly        pe_web.4            registry:develop-app    gtiDesaUbuDockerMaster03   Running             Running about an hour ago
```

And then add service name "web" to the stack name "pe" you specify "pe_web"

```
# docker service ps pe_web
ID                  NAME                IMAGE                  NODE                       DESIRED STATE       CURRENT STATE               ERROR               PORTS
u3tdzcxvh27z        pe_web.1            registry:develop-app   gtiDesaUbuDockerWorker01   Running             Running 6 days ago
p0ig9aoiljju         \_ pe_web.1        registry:develop-app   gtiDesaUbuDockerWorker01   Shutdown            Shutdown 6 days ago
tvlc28rq45ly        pe_web.4            registry:develop-app   gtiDesaUbuDockerMaster03   Running             Running about an hour ago
```


How this helps in some way in while not available...
Regards

**Comment by davidthornton** [comment]
Hi all, I know this is very tenuously an "issue" and perhaps a feature request, but since the transition to k8s as an orchestrator I get this

    $ docker service ls
    Error response from daemon: This node is not a swarm manager. Use "docker swarm init" or "docker swarm join" to connect this node to swarm and try again.

I came here looking for something like `docker stack logs <service name>` for use with my local k8s orchestrator. I know I can do

    $ docker logs -f <container name>

But it uses the ephemeral k8s allocated name. I know it's not much to complain about, but that to me proves more than anything what I consider to be the great state of Docker in 2k18. Well done team!

**Comment by manikmi** [comment]
I am creating a bunch of services on a swarm using the docker stack deploy --compose-file <>.. command. the problem if some of the services fail to start, they are not even shown up in the docker ps command on the respective docker machine. So, can't really get the error what might have happened? Can we have a log of docker stack deploy command?

**Comment by thaJeztah** [comment]
@manikmi if the service fails to start altogether, you may want to look at `docker service ps --no-trunc` to see why the containers failed to start; you can view logs with `docker service logs` but they may not be useful if the task failed.

**Comment by manikmi** [comment]
Thanks @theJeztah. It helped. Now I am facing another issue. the service gets started but fails very quickly. So, i cant see the issue in the service logs. Also, as I try to see the logs in the docker-machine using docker logs <container id>, the id is not found as it is getting restarted again and again. Any idea, how to check the logs?

**Comment by thaJeztah** [comment]
@manikmi if you scale down your service to 0 (`docker service scale <myservice>=0`), it will stop deploying the service but keep the last X failed tasks (and containers)

**Comment by masterBrog** [comment]
Elaborating on @dperny 's suggestion, as a work-around:

`docker stack services -q <stack> | xargs -L1 -I{} sh -c 'docker service logs -f {} &'`

These are spawned outside of shell so to stop them in a quick and dirty way `pkill -f 'docker service logs'` There's probably a better way but it worked for me

**Comment by martinelsner** [comment]
Not exactly the same as `docker-compose logs -f`, but I'm pretty happy using [multitail](https://www.vanheusden.com/multitail/) for this. It splits the console and shows one pane for every log. 
I've made a script like this:

```bash
#!/bin/bash -e
stack=$1
if [ -z "${stack}" ]; then
    echo 'Usage:'
    echo "$0 <stack>"
    exit 1
fi
# Get a list of the service names
services=$(docker stack services --format {{.Name}} ${stack})
# Sort the service names
services=$(echo "${services}" | sort)
# Create the command to run
command='multitail -s 2'
for service in ${services}; do
    command="${command} -l 'docker service logs -f ${service}'"
done
# Run the command
bash -c "${command}"
```
I put it in my ~/bin and call it like:
```bash
docker-stack-logs my-stack-name
```
`multitail` also has options to merge the outputs and a lot more. Just look at the [manpage](https://linux.die.net/man/1/multitail) and modify the script accordingly.

**Comment by zmarffy** [comment]
> ```shell
> #!/bin/bash -e
> stack=$1
> if [ -z "${stack}" ]; then
>     echo 'Usage:'
>     echo "$0 <stack>"
>     exit 1
> fi
> # Get a list of the service names
> services=$(docker stack services --format {{.Name}} ${stack})
> # Sort the service names
> services=$(echo "${services}" | sort)
> # Create the command to run
> command='multitail -s 2'
> for service in ${services}; do
>     command="${command} -l 'docker service logs -f ${service}'"
> done
> # Run the command
> bash -c "${command}"
> ```

Holy cow, this is really cool stuff. Good work. Not sure how a dumbed-down feature of this doesn't exist yet for `stack` though.

**Comment by gaving** [comment]
While workarounds are being thrown about, this seems a decent enough use case for [xpanes](https://github.com/greymd/tmux-xpanes):-

```bash
    docker stack services <stack-name> --format '{{.Name}}' | \
        xpanes -s -c "docker service logs -f {}"
```

**Comment by augnustin** [comment]
Up up up!

No resolution on this 3 years later?? OMG

Opening N terminals is not an acceptable workaround for me. And @Wolfium solution doesn't seem to work anymore.

```
docker service ps stack
$ no such service: stack
```

As far as I am concerned, I look at `docker ps -a` output, to check if some container are exited/restarted.

So the best I have is `watch -n 5 docker ps -a`.

But it would be much more convenient and logical to have all logs merged altogether.

What should we get from this inactivity? Is docker swarm un-maintained or something?

Cheers

**Comment by augnustin** [comment]
Digging [a little deeper](https://unix.stackexchange.com/questions/64736/combine-the-output-of-two-commands-in-bash), I went for the following script:

```
#!/bin/bash

SERVICES=$(docker stack services avril --format "{{.Name}}")

COMMAND="{ "
JOINER=""

for SERVICE in $SERVICES
do
  COMMAND="$COMMAND $JOINER docker service logs -f $SERVICE"
  JOINER="&"
done

COMMAND="$COMMAND; }"

eval $COMMAND
```

Executes: `{ docker service logs -f stack_service1 & docker service logs -f stack_service2; }`

> unless you make multiple calls (which is silly ...)

It may be silly but at least it is better than nothing. :P

**Comment by cedie712** [comment]
Been stumbled upon this issue lately. My first solution is to rely on ELK stack which i found nice and awesome but it requires more resources and I think it is an overkill for what I want to achieve.

**Comment by cedie712** [comment]
Been stumbled upon this issue lately. My first solution is to rely on ELK stack which i found nice and awesome but it requires more resources and I think it is an overkill for what I want to achieve.
So, I created my own work-around for this issue and here it is : https://github.com/cedie712/docker-swarm-centralized-logging. Hope it helps somebody who get into this problem.

**Comment by jahlives** [comment]
@augnustin 
thanks for your inspiration to "solve" that issue :+1: Took your code and changed it a bit to be more flexible: stackname and all options supported by `docker service logs` can be provided
```
#!/bin/bash

ARGS=''
SEARCH='Jobserver'
LASTPONG=$(date +%s)

# getting positional (aka last) argument as STACK
[[ "x${@:$#}" != 'x' && "x$(echo ${@:$#}|grep ^\-)" = 'x' ]] && STACK="${@:$#}"

# all other cli arguments go 1:1 to docker service logs
# used as docker service logs $ARGS $STACK
ARGS=${*%${!#}}

[[ -z ${STACK} || "x${STACK}" = 'x' ]] && echo "$0 stackname" && exit 1

SERVICES=$(docker stack services $STACK --format "{{.Name}}")
[[ "x$SERVICES" = 'x' || "$SERVICES" =~ ^Nothing ]] && exit 2

COMMAND="{ "
JOINER=""

for SERVICE in $SERVICES
do
  COMMAND="$COMMAND $JOINER docker service logs $ARGS $SERVICE"
  JOINER="&"
done

COMMAND="$COMMAND; }"

eval $COMMAND 2>&1 | grep --line-buffered '.*' | while read LOGLINE ; do
  [[ "${LOGLINE}" =~ $SEARCH ]] && LASTPONG=$(date +%s)
  [[ $(( $(date +%s) - $LASTPONG )) -gt 11 ]] && echo "$(TZ='UTC' date) ALERT"
done
```
last lines show how to pass the output to a while read loop which allows to act upon every new line.  Very handy for acting on certain logline near realtime. 
In sample above I check each line for $SEARCH and store a timestamp on $LASTPONG upon match. Then $LASTPONG can be compared to current time and trigger an alert action (in case above after 11s with no new line of $SEARCH)

**Comment by ghuzar** [comment]
> ```bash
> #!/bin/bash -e
> stack=$1
> if [ -z "${stack}" ]; then
>     echo 'Usage:'
>     echo "$0 <stack>"
>     exit 1
> fi
> # Get a list of the service names
> services=$(docker stack services --format {{.Name}} ${stack})
> # Sort the service names
> services=$(echo "${services}" | sort)
> # Create the command to run
> command='multitail -s 2'
> for service in ${services}; do
>     command="${command} -l 'docker service logs -f ${service}'"
> done
> # Run the command
> bash -c "${command}"
> ```

This approach could also be an option in case multitail is not available: https://stackoverflow.com/a/38308974/1555615

**Comment by mnahinkhan** [comment]
Another rather simple one:

`docker service ls -q | xargs -I _ sudo docker service logs _ | less`
